documentation/advanced/tutorial-i18n-localization.asciidoc

```diff
@@ -169,3 +169,146 @@ public class MyLocale extends Div {
     }
 }
 ----
+
+== Supporting Right-to-Left Mode
+
+Vaadin components have support for RTL languages. The components will work
+out-of-the-box on this mode, but to make your application handle this mode, a
+few changes are needed.
+
+On top of the last examples, let's say that your application is now also
+translated into a RTL Language, such as Arabic. Besides <<Provider sample for
+translation, providing the sample for the Arabic translation>>, at your main
+layout you can add a code like such:
+
+[source,java]
+----
+public class MainLayout extends VerticalLayout {
+
+    public MainLayout() {
+        // ...
+        final UI ui = UI.getCurrent();
+        if (ui.getLocale().getLanguage() == "ar") {
+            ui.setDirection(Direction.RIGHT_TO_LEFT);
+        }
+    }
+}
```

Comment:
As a protip, you can use the following to include the source code in the text, instead of copy-pasting it.

```asciidoc
[source, java]
----
include::../src/main/java/com/vaadin/flow/tutorial/advanced/RTLSources.java[tags=*, indent=0]
----
```

You should add a “tag” around the code block (using specially formatted comments) that you want to show, unless you want all of the code in the file to be shown. For example:

```java
@CodeFor("advanced/tutorial-i18n-localization.asciidoc")
 public class RTLSources {
// tag::rtl-app[]
     public class MainLayout extends VerticalLayout {

         public MainLayout() {
             // ...
             final UI ui = UI.getCurrent();
             if (ui.getLocale().getLanguage() == "ar") {
                 ui.setDirection(Direction.RIGHT_TO_LEFT);
             }
         }
     }
// end::rtl-app[]
 } 
```

[That’s a built-in AsciiDoc feature](https://asciidoctor.org/docs/user-manual/#include-partial), which I’m recommending all teams start using eventually, so that code is more easily verified to actually work and kept up-to-date in the written documentation.

Reply:
Well, that's nice to know. I haven't used AsciiDoc for a long time and, back then, nothing complex.